src/Background/Portal.vala

```diff
@@ -0,0 +1,162 @@
+
+    public Portal (DBusConnection connection) {
+        this.connection = connection;
+        notification_handler = new NotificationHandler (connection);
+        try {
+            desktop_integration = connection.get_proxy_sync ("org.pantheon.gala", "/org/pantheon/gala/DesktopInterface");
+            desktop_integration.running_applications_changed.connect (() => running_applications_changed ());
+        } catch {
+            warning ("Cannot connect to compositor, background portal working with reduced functionality.");
+        }
+    }
+
+    [DBus (name = "org.pantheon.gala.DesktopIntegration")]
+    public interface DesktopIntegration : Object {
+        public struct RunningApplications {
+            string app_id;
+            HashTable<string,Variant> details;
+        }
+
+        public signal void running_applications_changed ();
+        public abstract RunningApplications[] get_running_applications () throws DBusError, IOError;
+    }
+
+    private enum ApplicationState {
+        BACKGROUND,
+        RUNNING,
+        ACTIVE
+    }
+
+    public HashTable<string, Variant> get_app_state () throws DBusError, IOError {
+        if (desktop_integration == null) {
+            throw new DBusError.FAILED ("No connection to compositor.");
+        }
+
+        var apps = desktop_integration.get_running_applications ();
+        var results = new HashTable<string, Variant> (null, null);
+        foreach (var app in apps) {
+            var app_id = app.app_id;
+            if (app_id.has_suffix (".desktop")) {
+                app_id = app_id.slice (0, app_id.last_index_of_char ('.'));
+            }
+
+            results[app_id] = ApplicationState.RUNNING; //FIXME: Don't hardcode: needs implementation on the gala side
+        }
+
+        return results;
+    }
+
+    public async void notify_background (
+        ObjectPath handle,
+        string app_id,
+        string name,
+        out uint32 response,
+        out HashTable<string, Variant> results
+    ) throws DBusError, IOError {
+        response = 0; //Won't be used
+        var _results = new HashTable<string, Variant> (str_hash, str_equal);
+
+        var notification_request = notification_handler.send_notification (app_id, name);
+
+        if (notification_request == null) {
+            _results.set ("result", 2);
+            results = _results;
+            return;
+        }
+
+        notification_request.response.connect ((result) => {
+            _results.set ("result", result);
+            notify_background.callback ();
+        });
+
+        try {
+            notification_request.register_id = connection.register_object<NotificationRequest> (handle, notification_request);
+        } catch (Error e) {
+            critical ("Failed to export request object: %s", e.message);
+        }
+
+        yield;
+
+        connection.unregister_object (notification_request.register_id);
+        results = _results;
```

Comment:
register_id can be local. throw a DBusError.OBJECT_PATH_IN_USE if we fail to export the request.

```suggestion
        var notification_request = notification_handler.send_notification (app_id, name);
        if (notification_request == null) {
            response = 2;
            return;
        }

        uint register_id = 0;
        var _results = new HashTable<string, Variant> (str_hash, str_equal);
        var _response = 2;

        notification_request.response.connect ((result) => {
            if (result != NotifyBackgroundResult.CANCELLED) {
                _results["result"] = result;
                _response = 0;
            } else {
                _response = 1;
            }
            
            notify_background.callback ();
        });

        try {
            register_id = connection.register_object (handle, notification_request);
        } catch (Error e) {
            warning ("Failed to export request object: %s", e.message);
            throw new DBusError.OBJECT_PATH_IN_USE (e.message);
        }

        yield;

        connection.unregister_object (register_id);
        response = _response;
        results = _results;
```

Reply:
Why are we setting response here to different values? According to [documentation](https://docs.flatpak.org/en/latest/portal-api-reference.html#gdbus-method-org-freedesktop-impl-portal-Background.NotifyBackground) it shouldn't do anything as it states it's "not used" or is this best practice?

Reply:
it's not checked, but have meaning, 0 for success, 1 for cancelled and 2 for failed. the others portals only return the "result" key in case of a successful call, the frontend will default to ALLOW_ONCE in the absence of a result key.

Reply:
I'm not sure if I follow but I added it in :)